README.md
# Nsp is a Scientific Software Package. 

* It is based on a complete new rewrite of ScilabGtk
* The interpreter is written in C and objects with an internal class system.
* Gtk toolkit can be used from Nsp through a set of generated wrappers. The language bindings and class for Nsp are generated, the generator being based on the pygtk generator for python.
* It is modular (modular interpreter design, possible dynamic link of internal and external libraries).
* It should compile on Linux, MacOSX-macports, MacOSX-homebrew, Windows-Cygwin, Windows-Mingwin native Win32.
* A source version is available under CVS. Archive tarball, Fedora rpms, Debian package and win32 installer will be also provided.
* It is a GPL software.
* A Google group for Nsp is available at Tumbi Google Group.
* A bug report data base is available at Issues.

- Author(s):
  - Jean-Philippe Chancelier
  - Bruno Pinçon 
  - François Delebecque 
  - many others for used libraries
  



ChangeLog
2015-05-31  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* new_graphics is now the default
	* the class generator is now in ocaml (src/nspbind)
	* many other changes since last change log entry.

2012-01-08  Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* add glpk interface (linprog function).

2011-08-03  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* src/objects/serial.c: serialized object can be converted
	to base64 strings or matrix. This is used to implement a
	print as_read for serial objects.

2011-01-22  Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* add kron (.*.) for sparse (and a small test for kron in sparse-nsp.tst)

2011-01-08  Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* add intg_splin to integrate a spline or a sub-spline.

2010-08-09  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* scicos moved in toolboxes. remove old directories.

2010-04-11  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* src/gtksci/*: changes and addition in order to use
	gtksourceview for text editor widget

2010-01-16  Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* macros/statistics: add some options to mean, var, and std
	(st_deviation and variance tagged as obsolete), add median
	function and a test file.

2009-07-17  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* src/liboptim: optim library added.

2009-07-06  Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/libinteg/int3dnsp.c: add int3d function (together with its
	help and a test file)

2009-05-17  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* src/libportaudio added.

2008-11-18  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* src/libminpack added

2008-09-04  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/objects/MatOps.c (nsp_mat_complex): add complex
	function

2008-09-01  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/objects/MatOps.c (nsp_mat_angle): add
	angle function

2008-08-28  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* xexport can use cairo for exporting graphics even
	for non cairo graphic windows. pdf svg ps can be used
	png depends on the cairo/gtk version available. The
	exact list of available format should be added.

2008-08-21  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* diagset functions moved to set_diag method for
	matint objects.

2008-07-28  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* eye,ones,zeros: semantic has changed to follow Matlab rules.
	eye_new, ones_new, zeros_new are the same. eye_deprecated,
	ones_deprecated, zeros_deprecated follow old rules.


2008-07-18  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* GetActions added and improvement in GList GSlist management
	in gtk2/codegen.

2008-05-02  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* clear, clearglobal without arguments added for
	clearing all the variables of the current or global frame.

2008-05-01  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* *.sci: changes ones, zeros, eye to xxx_deprecated or
	xxx_new in sci, sce and tst files.

2008-04-11  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* macros/miscellaneous/bdiv_sp_m.sci: add a first
	version for backslash with sparse matrix.

2008-04-07  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* macros/miscellaneous/derivative.sci: add derivative
	function with its help page.
	* src/lapack/lapack.c: replace dgbcon and zgbcon with
	nsp_dgbcon and nsp_zgbcon (same calling sequence). They
	are faster but certainly less robust.

2008-04-03  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/libumfpack/umfpack.c add rcond method
	(nsp_umfpack_dlacon, nsp_umfpack_zlacon) for
	umfpack objects.

2008-03-30  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* src/type-test/codegen:
	generation of code for full_copy of objects
	which are by-reference objects.

2008-03-25  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* src/lapack/lapack.c: add the symmmetric and positive
	definite symmetric cases to bdiv.
	* src/lapack/lapack-IN.c: add interface for solve
	which uses the different bdiv functions. man added

2008-03-25  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/lapack/lapack.c (nsp_mat_bdiv_banded):
	add the banded case for bdiv. This comes with
	some utilities to detect lower and upper matrix
	bandwidths (for Mat and SpColMat too)
	* src/objects/SpColMatrix.c: a few utilities
	for sparse, 1/add triangular solving for sparse
	(interfaced with solve_tri but this is just to
	test), 2/ add scale rows and cols methods.
	* man/linalg/norm.tex
	add man page for norm

2008-03-17  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/objects/SpColMatrix.c (nsp_spcolmatrix_is_symetric):
	add issymetric test for sparse matrix.

2008-03-10  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/libinteg/integ-IN.c (make_triangles):
	int2d could now integrate directly on a rectangle
	(when X and Y arg of int2d are vectors of length 2
	a triangulation is computed), see int2d help page.
	* src/objet/SpColMatrix.c: rewrite sparse function
	(add 2 elements having the same (i,j) indices), write
	codes 1/ to compute frobenius norm (using lapack 's
	method of scaling) and 2/ to detect if a sparse
	is lower or upper triangular.

2008-03-06  Jean-Philippe Chancelier  <jpc@cermics.enpc.fr>

	* src/object/*.c:
	Storage of implicit vector in matrices added.
	A new method named as_index was also added to NspObject.
	as_index is implemented in NspMatrix, NspBMatrix ans NspIVect.
	as_index can use the implicit vector storage of
	matrices. matint.c and sparse matrices were changed
	to use as_index.

2008-03-03  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/libspmf/spmf-IN.c (int_nsp_pdf): add a pdf function
	(2 files are added which correspond to an adaptation
	for nsp of the Catherine Loader dbinom 's package)

2008-03-03  Pincon Bruno  <bruno.pincon@iecn.u-nancy.fr>

	* src/libinteg/integ-IN.c (intg_func): add the case
	of infinite interval for intg
	* src/libinteg/fromquadpack.f: source from the netlib
	quadpack used by nsp, adapted for nsp and cleaned


pvm3/src/pkt.h

/* Id: pkt.h,v 1.2 2008-05-31 06:23:45 jpc Exp $ */

/*
 *         PVM version 3.4:  Parallel Virtual Machine System
 *               University of Tennessee, Knoxville TN.
 *           Oak Ridge National Laboratory, Oak Ridge TN.
 *                   Emory University, Atlanta GA.
 *      Authors:  J. J. Dongarra, G. E. Fagg, M. Fischer
 *          G. A. Geist, J. A. Kohl, R. J. Manchek, P. Mucci,
 *         P. M. Papadopoulos, S. L. Scott, and V. S. Sunderam
 *                   (C) 1997 All Rights Reserved
 *
 *                              NOTICE
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby granted
 * provided that the above copyright notice appear in all copies and
 * that both the copyright notice and this permission notice appear in
 * supporting documentation.
 *
 * Neither the Institutions (Emory University, Oak Ridge National
 * Laboratory, and University of Tennessee) nor the Authors make any
 * representations about the suitability of this software for any
 * purpose.  This software is provided ``as is'' without express or
 * implied warranty.
 *
 * PVM version 3 was funded in part by the U.S. Department of Energy,
 * the National Science Foundation and the State of Tennessee.
 */

/*
 *	pkt.h
 *
 *	Packet buffers.
 *
 * : pkt.h,v $
 * Revision 1.1.1.1  2004/04/26 15:36:58  stochopt
 * Imported sources
 *
 * Revision 1.1.1.1  2003/11/14 13:02:08  stochopt
 * Imported files
 *
 * Revision 1.1.1.1  2003/02/18 09:26:05  stochopt
 * initial
 *
 * Revision 1.1.1.1  2003/02/17 11:25:32  stochopt
 * initial
 *
 * Revision 1.1.1.1  2003/02/03 08:42:12  stochopt
 * imported sources
 *
 * Revision 1.1.1.1  2003/01/06 09:04:12  stochopt
 * Imported sources
 *
 * Revision 1.1.1.1  2002/12/19 17:37:37  stochopt
 * start
 *
 * Revision 1.2  2002/10/14 14:37:50  chanceli
 * update
 *
 * Revision 1.6  1999/07/08 19:00:22  kohl
 * Fixed "Log" keyword placement.
 * 	- indent with " * " for new CVS.
 *
 * Revision 1.5  1997/06/25  22:09:17  pvmsrc
 * Markus adds his frigging name to the author list of
 * 	every file he ever looked at...
 *
 * Revision 1.4  1997/04/24  20:59:42  pvmsrc
 * added pk_rip structure member (used on mpps)
 *
 * Revision 1.3  1997/03/06  21:09:10  pvmsrc
 * 	- define __PKT_H__ so that multiple includes of pkt.h won't honk.
 *
 * Revision 1.2  1997/01/28  19:28:16  pvmsrc
 * New Copyright Notice & Authors.
 *
 * Revision 1.1  1996/09/23  23:43:22  pvmsrc
 * Initial revision
 *
 * Revision 1.3  1995/07/24  18:26:09  manchek
 * added message header fields and cpos, removed PF_STARTED
 *
 * Revision 1.2  1994/06/03  20:38:20  manchek
 * version 3.3.0
 *
 * Revision 1.1  1993/08/30  23:26:49  manchek
 * Initial revision
 *
 */

#ifndef __PKT_H__
#define __PKT_H__
struct pkt {
	struct pkt *pk_link;		/* queue or 0 */
	struct pkt *pk_rlink;
	struct pkt *pk_tlink;		/* scheduling queue or 0 */
	struct pkt *pk_trlink;
	char *pk_buf;				/* buffer or zero if master pkt */
	char *pk_dat;				/* data */
	int pk_max;					/* size of buffer */
	int pk_len;					/* length of data */
	int pk_src;					/* source tid */
	int pk_dst;					/* dest tid */
	int pk_flag;				/* fragment flags */
	int pk_enc;					/* message encoding */
	int pk_tag;					/* message type tag */
	int pk_ctx;					/* message context */
	int pk_wid;					/* message wait context id */
	int pk_crc;					/* message checksum */
	int pk_nrt;					/* retry count */
	int pk_seq;					/* seq num */
	int pk_ack;					/* ack num */
	int pk_rip;					/* received in place - used on mpps */
	struct hostd *pk_hostd;		/* receiving host */
	struct timeval pk_rtv;		/* time to retry */
	struct timeval pk_rta;		/* next-retry accumulator */
	struct timeval pk_rto;		/* total time spent on pkt */
	struct timeval pk_at;		/* time pkt first sent */
	char *pk_cpos;				/* write pointer for local protocol */
};

struct pkt *pk_new __ProtoGlarp__((int n));
void pk_free __ProtoGlarp__((struct pkt *p));
#endif
